proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
+   ...
 }
 ```
-The compiler may choose the following order for resolving the evidence that a `Monoid<Int>` exists in scope.
 
-1. Look in the most immediate scope for declarations of `with Monoid<Int>` in this case the function `addInts`
+### Type-Side Implementations
+
+This definition site poses additional complications when you consider multi-parameter typeclasses.
 
-This will compile because the responsibility of providing `Monoid<Int>` is passed into the callers of `addInts()`:
 ```kotlin
-extension object AddingInts {
-  fun addInts(with Monoid<Int>): Int = add(1, 2)
+package foo.collections
+
+extension interface Isomorphism<A, B> {
+   ...
 }
 ```
 
-2. Look in the immediately outher class/interface scope for declarations of `with Monoid<Int>` in this case the class `AddingInts`:
 ```kotlin
-extension class AddingInts(with Monoid<Int>) {
-  fun addInts(): Int = add(1, 2)
+package data.foo
+
+data class Foo(...)
+extension class<A> : Isomorphism<Foo, A> {
+   ...
 }
 ```
-This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `AddingInts()`
 
-3. Look in the import declarations for an explicitly imported instance that satisfies the constrain `Monoid<Int>`:
 ```kotlin
-import intext.IntMonoid
-extension object AddingInts {
-  fun addInts(): Int = add(1, 2)
+package data.bar
+
+data class Bar(...)
+extension class<A> : Isomorphism<A, Bar> {
+   ...
 }
 ```
-This will compile because the responsibility of providing `Monoid<Int>` is satisfied by `import intext.IntMonoid`
 
-4. Fail to compile if neither outer scopes nor explicit imports fail to provide evidence that there is a constrain satisfied by an instance in scope.
+The above instances are each defined alongside their respective type definitions and yet they clearly conflict with each other. We will also run into quandaries once we consider generic types. We can crib some prior art from Rust<sup>1</sup> to help us out here.
+
+To determine whether a typeclass definition is a valid type-side implementation we perform the following check:
+
+1. A "local type" is any type (but not typealias) defined in the current file (e.g. everything defined in `data.bar` if we're evaluating `data.bar`).
+2. A generic type parameter is "covered" by a type if it occurs within that type, e.g. `MyType` covers `T` in `MyType<T>` but not `Pair<T, MyType>`.
+3. Write out the parameters to the type class in order.
+4. The parameters must include a type defined in this file.
+5. Any generic type parameters must occur after the first instance of a local type or be covered by a local type.
+
+If a type class implementation meets these rules it is a valid type-side implementation.
+ 
+
+## Compile Resolution Rules
+
+When the compiler finds a call site invoking a function that has type class instances constraints declared with `with` as in the example below:
+
+Declaration: 
+```kotlin
+fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b)
```

Comment:
Can concrete instances be referenced directly? `with IntMonoid` or just the type class abstract definition?

Reply:
Since there can only be one implementation for a given type, they cannot be referred to by name. A typical strategy for types with multiple useful implementations is to provide wrapper types (e.g. `List` vs `ZipList`).

Reply:
makes sense